### PR TITLE
Bump Elasticsearch version to 9.0.0

### PR DESCRIPTION
**Pull request by mark-vieira**
This is the initial work to bump the Elasticsearch version to 9.0 in preparation for our next major version release.

### What this pull request includes

- The stack version has been bumped to 9.0.0.
- Backward compatibility testing logic has been updated to account for the new major version and branching changes. That is, once this PR is merged to `main`, the current `main` will be branched to `8.x`. Back compat testing will now checkout the `8.x` branch for testing that snapshot version.
- The minimum wire compatibility version has been updated to `8.16.0`. This is now determined dynamically based on known versions so no further changes should be necessary when `8.x` is bumped to `8.17`.
- Rest compatibility tests are running using tests sourced from the `8.x` branch. These are effectively identical to those in `main` so all existing rest compat overrides and transforms have been removed.
- Our rest compatibility plugins have been simplified as well. Task names no longer include the previous major version as there will only ever be one of them. This will make future major version bumps slightly easier.
- BWC test versions have been updated to remove all 7.x versions since there is now no level of backward compatibility with that version of Elasticsearch.
- The minimum index version has been bumped to `V_8_0_0` and all references **in tests** to index versions earlier to this have either been removed, or the test muted. This should mean that we can start wiping out production references to these versions, or replace them with placeholders, without affecting tests.
- Any tests that were explicitly testing upgrades or migrations from 7.x index versions have been removed or muted. If it was unclear whether the test was strictly related to 7.x (since it likely was using a 7.x version constant) it was muted and annotated with `@UpdateForV9` for later investigation.
- The EQL correctness tests have been muted since they rely on restoring a snapshot of test data. This snapshot unfortunately is from Elasticsearch 7.x, so cannot be restored in a 9.0 cluster. A new 9.0-compatible snapshot will need to be created for these tests to be re-enabled. See https://github.com/elastic/elasticsearch/issues/112572.

### What this pull request _does not_ include

- Production code references to pre-8.0 index versions have not yet been removed. This will be done in a follow up.
- The Lucene version remains 9.x. The upgrade to Lucene 10 is a separate effort.
- The minimum transport version remains `V_7_17_0`. This will be bumped in a follow up.
- The REST API version remains `V_8`. A constant has been added for `V_9` but is not used. Also, the bootstrap check to ensure the current REST API version matches the current major version has been disabled. This will need to be re-enabled once the REST API version is incremented.
- No functional changes. For all intents and purposes this behaves exactly like 8.x, with the exception of the removal of support for pre-8.0 indexes. Anything deprecated for removal in 9.0 still remains. This will be done in a follow up.
~Native ML code is still version 8.16.0. We do not yet have 9.0 snapshot builds of ml-cpp, so this will need to be updated in a follow up.~ (done in this PR)

### CODE REVIEW

File: .ci/scripts/run-gradle.sh [hunk 1]
```diff
@@ -37,4 +37,4 @@ if [[ "$MAX_WORKERS" == "0" ]]; then
 fi
 
 set -e
-$GRADLEW -S --max-workers=$MAX_WORKERS $@
+$GRADLEW -S --max-workers=$MAX_WORKERS --continue $@
```
Comment (brianseeders) [inline .ci/scripts/run-gradle.sh:40]: Did you mean to leave this here as well?

Comment (mark-vieira) [reply .ci/scripts/run-gradle.sh:40]: I think I'll remove it from this PR and we can have a separate discussion as to whether we should do this across CI in general.

**Comment by brianseeders** [comment]
@elasticmachine merge upstream

**Review by jakelandis — APPROVED** [review]
LGTM

**Review by thecoop — APPROVED** [review]

File: build-tools-internal/src/main/java/org/elasticsearch/gradle/internal/BwcVersions.java [hunk 2]
```diff
@@ -286,7 +266,17 @@ public List<Version> getUnreleasedWireCompatible() {
     }
 
     public Version getMinimumWireCompatibleVersion() {
-        return MINIMUM_WIRE_COMPATIBLE_VERSION;
+        // Determine minimum wire compatible version from list of known versions.
+        // Current BWC policy states the minimum wire compatible version is the last minor release or the previous major version.
```
Comment (thecoop) [inline build-tools-internal/src/main/java/org/elasticsearch/gradle/internal/BwcVersions.java:270]: typo or -> of

File: build-tools-internal/src/main/java/org/elasticsearch/gradle/internal/test/rest/compat/compat/AbstractYamlRestCompatTestPlugin.java [hunk 3]
```diff
@@ -89,10 +89,19 @@ public void apply(Project project) {
         SourceSet yamlTestSourceSet = sourceSets.getByName(YamlRestTestPlugin.YAML_REST_TEST);
         GradleUtils.extendSourceSet(project, YamlRestTestPlugin.YAML_REST_TEST, SOURCE_SET_NAME);
 
+        // determine the previous rest compatibility version and BWC project path
+        int currentMajor = BuildParams.getBwcVersions().getCurrentVersion().getMajor();
+        Version lastMinor = BuildParams.getBwcVersions()
+            .getUnreleased()
+            .stream()
+            .filter(v -> v.getMajor() == currentMajor - 1)
+            .min(Comparator.reverseOrder())
```
Comment (thecoop) [inline build-tools-internal/src/main/java/org/elasticsearch/gradle/internal/test/rest/compat/compat/AbstractYamlRestCompatTestPlugin.java:98]: `max(Comparator.naturalOrder())` is more understandable

File: build-tools-internal/src/test/groovy/org/elasticsearch/gradle/internal/BwcVersionsSpec.groovy [hunk 4]
```diff
@@ -42,11 +39,12 @@ class BwcVersionsSpec extends Specification {
         unreleased == [
             (v('7.16.2')): new UnreleasedVersionInfo(v('7.16.2'), '7.16', ':distribution:bwc:bugfix'),
             (v('7.17.0')): new UnreleasedVersionInfo(v('7.17.0'), '7.17', ':distribution:bwc:staged'),
-            (v('8.0.0')): new UnreleasedVersionInfo(v('8.0.0'), '8.0', ':distribution:bwc:minor'),
+            (v('8.0.0')): new UnreleasedVersionInfo(v('8.0.0'), '8.x', ':distribution:bwc:minor'),
             (v('8.1.0')): new UnreleasedVersionInfo(v('8.1.0'), 'main', ':distribution')
         ]
         bwc.wireCompatible == [v('7.17.0'), v('8.0.0'), v('8.1.0')]
-        bwc.indexCompatible == osFiltered([v('7.14.0'), v('7.14.1'), v('7.14.2'), v('7.15.0'), v('7.15.1'), v('7.15.2'), v('7.16.0'), v('7.16.1'), v('7.16.2'), v('7.17.0'), v('8.0.0'), v('8.1.0')])
+        bwc.indexCompatible == [v('7.14.0'), v('7.14.1'), v('7.14.2'), v('7.15.0'), v('7.15.1'), v('7.15.2'), v('7.16.0'), v('7.16.1'), v('7.16.2'), v('7.17.0'), v('8.0.0'), v('8.1.0')]
```
Comment (thecoop) [inline build-tools-internal/src/test/groovy/org/elasticsearch/gradle/internal/BwcVersionsSpec.groovy:46]: I'm not sure what's going on here, but I assume it's deliberate that indexCompatible /minWireCompatible is still referring to v7 versions

**Comment by brianseeders** [comment]
@elasticmachine merge upstream

**Comment by jakelandis** [comment]
@elasticmachine merge upstream

**Comment by jakelandis** [comment]
@elasticmachine merge upstream

**Pull request merged**